### PR TITLE
chore(select): replacing reach lib with react-select

### PR DESCRIPTION
**Pull request by bdebon**
CleanShot 2022-09-07 at 18 39 34@2x

- [x] rewrite tests
- [x] find a way to set value of react-select during test
- [x] fix accessibility of inputs: focus / labelUp / valid state
- [x] remove reach
- [x] make sure bug is no longer present on both chrome and firefox

https://user-images.githubusercontent.com/6163954/188933572-8896f4b4-da89-48df-90d5-266a7862561c.mp4

### CODE REVIEW

**Comment by evoxmusic** [comment]
A preview environment was automatically created via Qovery.  
Click on the link below to follow its deployment and use it.  
:point_right: [[PR] staging - Chore(select): replacing reach lib with react-select - 2022-09-07T16:41:06Z](https://console.qovery.com/platform/organization/141c07c8-0dd9-4623-983b-3fdd61867255/projects/977bd069-20f7-4b24-bd95-28b44174cf43/environments/7416d76f-ed55-473b-8998-1cb898fa56f2/applications)

File: libs/pages/application/src/lib/ui/crud-environment-variable-modal/crud-environment-variable-modal.tsx [hunk 1]
```diff
@@ -75,8 +75,8 @@ export function CrudEnvironmentVariableModal(props: CrudEnvironmentVariableModal
           render={({ field, fieldState: { error } }) => (
             <InputSelect
               className="mb-6"
-              portal={false}
-              items={props.availableScopes.map((s) => ({ value: s, label: s.toLowerCase() }))}
+              portal={true}
```
Comment (RemiBonnet) [inline libs/pages/application/src/lib/ui/crud-environment-variable-modal/crud-environment-variable-modal.tsx:78]: When an props is true you can directly add `portal` without `={true}`

File: libs/pages/application/src/lib/ui/crud-environment-variable-modal/crud-environment-variable-modal.tsx [hunk 2]
```diff
@@ -75,8 +75,8 @@ export function CrudEnvironmentVariableModal(props: CrudEnvironmentVariableModal
...
+              options={props.availableScopes.map((s) => ({ value: s, label: s.toLowerCase() }))}
```
Comment (RemiBonnet) [inline libs/pages/application/src/lib/ui/crud-environment-variable-modal/crud-environment-variable-modal.tsx:79]: Cool to rename it with `options` ! 👍

File: libs/pages/application/src/lib/ui/git-repository-settings/git-repository-settings.tsx [hunk 3]
```diff
@@ -69,12 +69,12 @@ export function GitRepositorySettings(props: GitRepositorySettingsProps) {
...
                 label="Repository"
                 className="mb-3"
-                items={repositories}
+                options={repositories}
                 onChange={field.onChange}
                 value={field.value}
                 error={error?.message}
                 disabled={gitDisabled}
-                search
+                isSearchable={true}
```
Comment (RemiBonnet) [inline libs/pages/application/src/lib/ui/git-repository-settings/git-repository-settings.tsx:77]: Same here

File: libs/pages/environments/src/lib/feature/create-clone-environment-modal-feature/create-clone-environment-modal-feature.spec.tsx [hunk 4]
```diff
@@ -112,8 +116,12 @@ describe('CreateCloneEnvironmentModalFeature', () => {
         fireEvent.input(inputs[1], { target: { value: 'test' } })
       })
 
-      await changeSelectValueByLabel(baseElement, 'input-select-cluster', mockClusters[2].name)
-      await changeSelectValueByLabel(baseElement, 'input-select-mode', 'Staging')
+      await act(() => {
+        selectEvent.select(getByLabelText(baseElement, 'Cluster'), mockClusters[2].name, {
```
Comment (RemiBonnet) [inline libs/pages/environments/src/lib/feature/create-clone-environment-modal-feature/create-clone-environment-modal-feature.spec.tsx:120]: 💯

File: libs/shared/ui/src/lib/components/inputs/input-select-small/input-select-small.tsx [hunk 5]
```diff
@@ -42,7 +41,7 @@ export function InputSelectSmall(props: InputSelectSmallProps) {
         data-testid={dataTestId || 'input-select-small'}
         name={name}
         value={value}
-        className={`input input__select--small ${isValid ? 'input--focused' : ''}`}
+        className={`input input__select--small`}
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/components/inputs/input-select-small/input-select-small.tsx:44]: Can be `className="input input__select--small"`

File: libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx [hunk 6]
```diff
@@ -1,125 +1,170 @@
...
+    isClearable = false,
   } = props
+  const [focused, setFocused] = useState(false)
+  const [selectedItems, setSelectedItems] = useState<MultiValue<Value> | SingleValue<Value>>([])
+  const [selectedValue, setSelectedValue] = useState<string | string[]>([])
 
-  const [selectedValue, setSelectedValue] = useState(value && items.find((item) => item.value === value))
-  const hasError = error && error.length > 0 ? 'input--error' : ''
+  const hasFocus = focused
+  const hasError = error ? 'input--select-multiple--error' : ''
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx:48]: I see on this component a mix between `input--select` and `input--select-multiple`, it's possible to remove the `input-select-multiple` or add it when you need a specific class for multiple?

File: libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx [hunk 7]
```diff
@@ -1,125 +1,170 @@
...
+  }
 
   useEffect(() => {
-    setCurrentItems(items)
-  }, [items])
+    const items = options.filter((option) => {
+      if (isMulti) {
+        return (value as string[])?.includes(option.value)
+      } else {
+        if (Object.prototype.toString.call(value) === '[object Array]') {
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx:70]: Does it work like `type of` ? (maybe more simple to use type of)

File: libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx [hunk 8]
```diff
@@ -1,125 +1,170 @@
...
+          components={{
+            Option,
+            MultiValue,
+            SingleValue,
+          }}
+          name={label}
+          inputId={label}
+          closeMenuOnSelect={!isMulti}
+          onChange={handleChange}
+          classNamePrefix="input--select-multiple"
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx:152]: Here we didn't have this prefix if we use the input--select without multiple

File: libs/shared/ui/src/lib/styles/components/input.scss [hunk 9]
```diff
@@ -51,11 +63,16 @@
...
   @apply border-error-500;
   box-shadow: 0 2px 2px rgba(0, 0, 0, 0.05);
+
   &:hover {
     @apply border-error-500;
   }
 }
 
+.input--error.input--focused {
+  outline: 1px solid theme('colors.error.500') !important;
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/styles/components/input.scss:73]: I'm not a big fan of using a lot of !important, is it really necessary here?

File: libs/shared/ui/src/lib/styles/components/input.scss [hunk 10]
```diff
@@ -29,6 +34,13 @@
 
 .input--focused,
 .input__button--focused {
+  outline: 1px solid theme('colors.brand.500') !important;
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/styles/components/input.scss:37]: !important is necessary ?

File: libs/shared/ui/src/lib/styles/components/input.scss [hunk 11]
```diff
@@ -6,16 +6,21 @@
...
 
   &:hover:not(.input--error),
   &.input--focused:not(.input--error),
   .input__button--focused {
     @apply border-brand-500;
     box-shadow: 0 2px 2px rgba(0, 0, 0, 0.05);
+
+    input,
+    textarea {
+      outline: none !important;
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/styles/components/input.scss:22]: !important is necessary ?

File: libs/shared/ui/src/lib/styles/components/select-multiple.scss [hunk 12]
```diff
@@ -15,9 +15,14 @@
 .input--select-multiple__value-container {
   padding: 0 !important;
   cursor: pointer;
+  display: flex !important;
```
Comment (RemiBonnet) [inline libs/shared/ui/src/lib/styles/components/select-multiple.scss:18]: To example here, I understand why you use the !important we have default class by the react-select library

**Review by RemiBonnet — COMMENTED** [review]
Well done @bdebon ! 💯 

- Can we remove the duration on the outline ? I think it should be better because it's a bit little "laggy" with the duration 

- I think we have an mistake with a mix of class for multiple and class without multiple when we have a classic select, can you harmonize it ?

Thanks !!

File: libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx (hunk 7 above)
Comment (bdebon) [reply libs/shared/ui/src/lib/components/inputs/input-select/input-select.tsx:70]: I changed fo `Array.isArray(value)` https://stackoverflow.com/questions/4775722/how-can-i-check-if-an-object-is-an-array

**Pull request merged**